internal/fsutil/direct_io_reader.go

```diff
@@ -0,0 +1,98 @@
+package fsutil
+
+import (
+	"errors"
+	"io"
+	"os"
+	"sync"
+	"syscall"
+
+	"github.com/ncw/directio"
+	"github.com/spf13/viper"
+
+	conf "github.com/wal-g/wal-g/internal/config"
+)
+
+const directIOBlockCount = 32
+
+type reader struct {
+	mu           *sync.Mutex
+	fd           *os.File
+	buff         []byte
+	buffOffset   int
+	alignedBlock []byte
+}
+
+// OpenReadOnlyMayBeDirectIO returns read-only io.ReadSeekCloser.
+func OpenReadOnlyMayBeDirectIO(path string) (io.ReadSeekCloser, error) {
+	if viper.GetBool(conf.DirectIO) {
+		return NewDirectIOReadSeekCloser(path, syscall.O_RDONLY, 0)
+	}
+	return os.OpenFile(path, os.O_RDONLY, 0)
+}
+
+// NewDirectIOReadSeekCloser returns io.ReadSeekCloser.
+func NewDirectIOReadSeekCloser(path string, flag int, perm os.FileMode) (io.ReadSeekCloser, error) {
+	in, errOpen := directio.OpenFile(path, flag, perm)
+	if errOpen != nil {
+		return nil, errOpen
+	}
+	return &reader{
+		mu:           &sync.Mutex{},
+		fd:           in,
+		buff:         nil,
+		alignedBlock: directio.AlignedBlock(directIOBlockCount * directio.BlockSize),
+	}, nil
+}
+
+func (r *reader) readBuff() error {
+	if n, err := io.ReadFull(r.fd, r.alignedBlock); err != nil {
+		r.buff = append(r.buff[r.buffOffset:], r.alignedBlock[0:n]...)
+		r.buffOffset = 0
+		if errors.Is(err, io.ErrUnexpectedEOF) {
+			err = io.EOF
+		}
+		return err
+	}
+	r.buff = append(r.buff[r.buffOffset:], r.alignedBlock...)
+	r.buffOffset = 0
+	return nil
+}
+
+func (r *reader) copyBuff(p []byte) int {
+	n := copy(p, r.buff[r.buffOffset:])
+	r.buffOffset += n
+	return n
+}
+
+func (r *reader) Seek(offset int64, whence int) (int64, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.buffOffset = 0
+	r.buff = nil
+	r.alignedBlock = make([]byte, len(r.alignedBlock))
```

Comment:
should we use `directio.AlignedBlock()` here as well (or just zero all values in the original slice)?